File: examples/anthropic-managed-agents/python/anthropic_managed_agents_e2b/worker_runtime.py
from __future__ import annotations

import asyncio
import logging
import os

from anthropic import AsyncAnthropic

WORKDIR = "/mnt/session"


def max_idle_seconds() -> float | None:
    raw = os.environ.get("WORKER_MAX_IDLE_SECONDS", "30")
    if raw.lower() in {"", "none", "null"}:
        return None
    return float(raw)


def run_seconds() -> float | None:
    raw = os.environ.get("WORKER_RUN_SECONDS", "180")
    if raw.lower() in {"", "none", "null"}:
        return None
    return float(raw)


async def run_worker() -> None:
    logging.basicConfig(level=os.environ.get("LOG_LEVEL", "INFO"))
    logger = logging.getLogger(__name__)

    environment_id = os.environ["ANTHROPIC_ENVIRONMENT_ID"]
    environment_key = os.environ["ANTHROPIC_ENVIRONMENT_KEY"]
    max_run_seconds = run_seconds()

    async with AsyncAnthropic(auth_token=environment_key) as client:
        worker = client.beta.environments.work.worker(
            environment_id=environment_id,
            environment_key=environment_key,
            workdir=WORKDIR,
            max_idle=max_idle_seconds(),
        )
        runner = (
            worker.handle_item(
                work_id=os.environ.get("ANTHROPIC_WORK_ID"),
                environment_id=environment_id,
                session_id=os.environ.get("ANTHROPIC_SESSION_ID"),
                environment_key=environment_key,
            )
            if os.environ.get("ANTHROPIC_WORK_ID") or os.environ.get("ANTHROPIC_SESSION_ID")
            else worker.run()
        )
        if max_run_seconds is None:
            await runner
            return

        try:
            await asyncio.wait_for(runner, timeout=max_run_seconds)
        except TimeoutError:
            logger.info("worker reached WORKER_RUN_SECONDS=%s; exiting", max_run_seconds)


def main() -> None:
    asyncio.run(run_worker())

File: examples/anthropic-managed-agents/javascript/src/worker-runtime.ts
import Anthropic from "@anthropic-ai/sdk";
import { type LogLevel } from "@anthropic-ai/sdk/client";

const WORKDIR = "/mnt/session";
const LOG_LEVELS = new Set(["off", "error", "warn", "info", "debug"]);

function maxIdleMs() {
  const raw = process.env.WORKER_MAX_IDLE_SECONDS ?? "30";
  if (["", "none", "null"].includes(raw.toLowerCase())) {
    return undefined;
  }
  return Number(raw) * 1000;
}

function runMs() {
  const raw = process.env.WORKER_RUN_SECONDS ?? "180";
  if (["", "none", "null"].includes(raw.toLowerCase())) {
    return undefined;
  }
  return Number(raw) * 1000;
}

function logLevel(): LogLevel {
  const raw = (process.env.LOG_LEVEL ?? "info").toLowerCase();
  return (LOG_LEVELS.has(raw) ? raw : "info") as LogLevel;
}

async function main() {
  const environmentId = process.env.ANTHROPIC_ENVIRONMENT_ID;
  const environmentKey = process.env.ANTHROPIC_ENVIRONMENT_KEY;

  if (!environmentId) {
    throw new Error("ANTHROPIC_ENVIRONMENT_ID is required");
  }
  if (!environmentKey) {
    throw new Error("ANTHROPIC_ENVIRONMENT_KEY is required");
  }

  const client = new Anthropic({
    authToken: environmentKey,
    logger: console,
    logLevel: logLevel(),
  });
  const controller = new AbortController();
  const maxRunMs = runMs();
  const timer =
    maxRunMs === undefined ? undefined : setTimeout(() => controller.abort(), maxRunMs);

  try {
    const worker = client.beta.environments.work.worker({
      environmentId,
      environmentKey,
      workdir: WORKDIR,
      maxIdleMs: maxIdleMs(),
      signal: controller.signal,
    });
    if (process.env.ANTHROPIC_WORK_ID || process.env.ANTHROPIC_SESSION_ID) {
      await worker.handleItem({
        workId: process.env.ANTHROPIC_WORK_ID,
        environmentId,
        sessionId: process.env.ANTHROPIC_SESSION_ID,
        environmentKey,
        signal: controller.signal,
      });
    } else {
      await worker.run();
    }
  } finally {
    if (timer) {
      clearTimeout(timer);
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
